### ISSUE TITLE
Remove colour for metavars optional square brackets in usage

### ISSUE THREAD

**Issue by atteggiani** [hamdanal/rich-argparse#127]
### Description
Currently the usage text is formatted so that all the text in the metavar string gets coloured, including any spaces and square brackets to indicate optional metavars (I will refer to them as _optional square brackets_). 

#### Example output
This is an example of a current usage coloured output.
<img width="1014" alt="Screenshot 2024-06-28 at 1 36 27 PM" src="https://github.com/hamdanal/rich-argparse/assets/30042436/6988d401-6e7c-4bf8-93ee-d1bbf24a5213">

This is the same usage output but a background colour has been added to the style for both `argparse.args` and `argparse.metavars` to better illustrate the behaviour.
<img width="1014" alt="Screenshot 2024-06-28 at 2 26 43 PM" src="https://github.com/hamdanal/rich-argparse/assets/30042436/54d7052f-a9c8-4d93-bbaf-dfeadc5eaeee">

### Suggestion for improvement
I am raising this issue to scope for interest in changing the behaviour of `rich-argparse` to only colour actual metavars in the usage (leaving spaces and _optional square brackets_ with the default colour).
The usage output with the new behaviour would look like this: 
<img width="1014" alt="Screenshot 2024-06-28 at 1 36 38 PM" src="https://github.com/hamdanal/rich-argparse/assets/30042436/e2cc7bba-fb68-4292-b68f-4326e06c1108">
And the same usage output with the background colours:
<img width="1013" alt="Screenshot 2024-06-28 at 2 26 26 PM" src="https://github.com/hamdanal/rich-argparse/assets/30042436/e9b150e8-f37a-446e-8e60-507ab4730fb9">

This new behaviour would not break/change any current implementation apart from the usage output itself.
A full working implementation with test and 100% coverage is available in the [`usage_highlight_strict_metavars`](https://github.com/atteggiani/rich-argparse/tree/usage_highlight_strict_metavars) branch of my forked `rich-argparse` repo.

If this is something of interest, I can open a PR.

Thank you for the amazing module @hamdanal! 🚀

**Comment by hamdanal** [comment]
I don't mind this change, I am actually neutral.
Do you think the proposed change makes the text look better? I'd not expect a lot of people are setting a background color anyway but maybe not coloring the brackets would be welcome by users.

 I do think however that if we change how the metavar is colored in the usage we should also consider changing it in the help string as well so that both remain consistent.

I'll have to take a look at our metavar handling to see if there is room for improvement without adding too much complexity to the code. Feel free to make a PR and I can take a look.

**Comment by atteggiani** [comment]
> I don't mind this change, I am actually neutral. Do you think the proposed change makes the text look better? I'd not expect a lot of people are setting a background color anyway but maybe not coloring the brackets would be welcome by users.
> 
Yeah I also would not expect people to set a background color, my example was only to better illustrate the behaviour. 
Anyway, I think overall the text would look better if we highlight only the metavars without spaces or brackets.

> I do think however that if we change how the metavar is colored in the usage we should also consider changing it in the help string as well so that both remain consistent.
> 
> I'll have to take a look at our metavar handling to see if there is room for improvement without adding too much complexity to the code. Feel free to make a PR and I can take a look.
Yeah that makes sense. 
I'll try to figure that out and open a PR in case I manage.

**Comment by hamdanal** [comment]
I was looking at your other PR and it got me thinking that both this issue and the other one you fix in your PR can be fixed if we have a finer control on metavar creation. This way we avoid using regular expressions and we avoid duplicating the work for usage metavars and help text metavars. I came up with this patch (not tested yet):

<details><summary>Details</summary>
<p>

```patch
diff --git a/rich_argparse/__init__.py b/rich_argparse/__init__.py
index 69d4552..b0d9676 100644
--- a/rich_argparse/__init__.py
+++ b/rich_argparse/__init__.py
@@ -305,36 +305,84 @@ class RichHelpFormatter(argparse.HelpFormatter):
                 usage = action.option_strings[0]
             start, end = find_span(usage)
             yield r.Span(start, end, "argparse.args")
-            if action.nargs != 0:
-                metavar = self._format_args(action, self._get_default_metavar_for_optional(action))
-                start, end = find_span(metavar)
-                yield r.Span(start, end, "argparse.metavar")
             pos = end + 1
-        for action in positionals:  # positionals come at the end
-            metavar = self._get_default_metavar_for_positional(action)
-            metavar_tuple = self._metavar_formatter(action, metavar)(1)
-            usage = metavar_tuple[0]
-            if isinstance(action.nargs, int):
-                nargs = action.nargs
-            elif action.nargs in (argparse.REMAINDER, argparse.SUPPRESS):
-                nargs = 0
-            elif action.nargs in (None, argparse.OPTIONAL, argparse.PARSER):
-                nargs = 1
-            elif action.nargs == argparse.ZERO_OR_MORE:
-                if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
-                    nargs = 2 if len(metavar_tuple) == 2 else 1
-                else:  # pragma: <3.9 cover
-                    nargs = 2
-            elif action.nargs == argparse.ONE_OR_MORE:
-                nargs = 2
-            else:  # pragma: no cover
-                # unknown nargs, fallback to coloring the whole thing
-                usage = self._format_args(action, metavar)
-                nargs = 1
-            for _ in range(nargs):
-                start, end = find_span(usage)
-                yield r.Span(start, end, "argparse.args")
+            if action.nargs != 0:
+                default_metavar = self._get_default_metavar_for_optional(action)
+                for metavar_part, colorize in self._rich_metavar_parts(action, default_metavar):
+                    start, end = find_span(metavar_part)
+                    if colorize:
+                        yield r.Span(start, end, "argparse.metavar")
+                    pos = end
                 pos = end + 1
+        for action in positionals:  # positionals come at the end
+            default_metavar = self._get_default_metavar_for_positional(action)
+            for metavar_part, colorize in self._rich_metavar_parts(action, default_metavar):
+                start, end = find_span(metavar_part)
+                if colorize:
+                    yield r.Span(start, end, "argparse.args")
+                pos = end
+            pos = end + 1
+
+    def _rich_metavar_parts(
+        self, action: Action, default_metavar: str
+    ) -> Iterator[tuple[str, bool]]:
+        get_metavar = self._metavar_formatter(action, default_metavar)
+        # similar to self._format_args but yields (part, colorize) of the metavar
+        if action.nargs is None:
+            # '%s' % get_metavar(1)
+            yield "%s" % get_metavar(1), True  # noqa: UP031
+        elif action.nargs == argparse.OPTIONAL:
+            # '[%s]' % get_metavar(1)
+            yield "[", False
+            yield "%s" % get_metavar(1), True  # noqa: UP031
+            yield "]", False
+        elif action.nargs == argparse.ZERO_OR_MORE:
+            metavar = get_metavar(1)
+            if len(metavar) == 2 or sys.version_info < (3, 9):
+                # '[%s [%s ...]]' % metavar
+                yield "[", False
+                yield "%s" % metavar[0], True  # noqa: UP031
+                yield " [", False
+                yield "%s" % metavar[1], True  # noqa: UP031
+                yield " ...", True
+                yield "]]", False
+            else:
+                # '[%s ...]' % metavar
+                yield "[", False
+                yield "%s" % metavar, True  # noqa: UP031
+                yield " ...", True
+                yield "]", False
+        elif action.nargs == argparse.ONE_OR_MORE:
+            # '%s [%s ...]' % get_metavar(2)
+            metavar = get_metavar(2)
+            yield "%s" % metavar[0], True  # noqa: UP031
+            yield " [", False
+            yield "%s" % metavar[1], True  # noqa: UP031
+            yield " ...", True
+            yield "]", False
+        elif action.nargs == argparse.REMAINDER:
+            # '...'
+            yield "...", True
+        elif action.nargs == argparse.PARSER:
+            # '%s ...' % get_metavar(1)
+            yield "%s" % get_metavar(1), True  # noqa: UP031
+            yield " ...", True
+        elif action.nargs == argparse.SUPPRESS:
+            # ''
+            yield "", False
+        else:
+            try:
+                list(range(action.nargs))  # type: ignore[arg-type]
+            except TypeError:
+                raise ValueError("invalid nargs value") from None
+            metavar = get_metavar(action.nargs)  # type: ignore[arg-type]
+            first = True
+            for met in metavar:
+                if first:
+                    first = False
+                else:
+                    yield " ", False
+                yield "%s" % met, True  # noqa: UP031
 
     def _rich_whitespace_sub(self, text: r.Text) -> r.Text:
         # do this `self._whitespace_matcher.sub(' ', text).strip()` but text is Text
@@ -433,8 +481,10 @@ class RichHelpFormatter(argparse.HelpFormatter):
             )
             if action.nargs != 0:
                 default = self._get_default_metavar_for_optional(action)
-                args_string = self._format_args(action, default)
-                action_header.append(" ").append(args_string, style="argparse.metavar")
+                action_header.append(" ")
+                for metavar_part, colorize in self._rich_metavar_parts(action, default):
+                    style = "argparse.metavar" if colorize else None
+                    action_header.append(metavar_part, style=style)
             return action_header
 
     def _rich_split_lines(self, text: r.Text, width: int) -> r.Lines:
```

</p>
</details> 

Is this what you want? Feel free to use this patch in your PR if it is what you want.

**Comment by atteggiani** [comment]
It makes a lot of sense to process the metavars creation only once and use it for both usage text and further specification text in the help. 
Your solution seems very good! 

Very few modifications I applied are:

1. Separated the spaces and ellipsis within the same metavar (so the space is not coloured).
2. Reformatted the multiple `yield` statements using one single `yield from (...)` 

<details>

```diff
diff --git a/rich_argparse/__init__.py b/rich_argparse/__init__.py
index 5f03ae3..62fbb4d 100644
--- a/rich_argparse/__init__.py
+++ b/rich_argparse/__init__.py
@@ -332,40 +332,54 @@ class RichHelpFormatter(argparse.HelpFormatter):
             yield "%s" % get_metavar(1), True  # noqa: UP031
         elif action.nargs == argparse.OPTIONAL:
             # '[%s]' % get_metavar(1)
-            yield "[", False
-            yield "%s" % get_metavar(1), True  # noqa: UP031
-            yield "]", False
+            yield from (
+                ("[", False),
+                ("%s" % get_metavar(1), True),  # noqa: UP031
+                ("]", False),
+            )
         elif action.nargs == argparse.ZERO_OR_MORE:
             metavar = get_metavar(1)
             if len(metavar) == 2 or sys.version_info < (3, 9):
                 # '[%s [%s ...]]' % metavar
-                yield "[", False
-                yield "%s" % metavar[0], True  # noqa: UP031
-                yield " [", False
-                yield "%s" % metavar[1], True  # noqa: UP031
-                yield " ...", True
-                yield "]]", False
+                yield from (
+                    ("[", False),
+                    ("%s" % metavar[0], True),  # noqa: UP031
+                    (" [", False),
+                    ("%s" % metavar[1], True),  # noqa: UP031
+                    (" ", False),
+                    ("...", True),
+                    ("]]", False),
+                )
             else:
                 # '[%s ...]' % metavar
-                yield "[", False
-                yield "%s" % metavar, True  # noqa: UP031
-                yield " ...", True
-                yield "]", False
+                yield from (
+                    ("[", False),
+                    ("%s" % metavar, True),  # noqa: UP031
+                    (" ", False),
+                    ("...", True),
+                    ("]", False),
+                )
         elif action.nargs == argparse.ONE_OR_MORE:
             # '%s [%s ...]' % get_metavar(2)
             metavar = get_metavar(2)
-            yield "%s" % metavar[0], True  # noqa: UP031
-            yield " [", False
-            yield "%s" % metavar[1], True  # noqa: UP031
-            yield " ...", True
-            yield "]", False
+            yield from (
+                ("%s" % metavar[0], True),  # noqa: UP031
+                (" [", False),
+                ("%s" % metavar[1], True),  # noqa: UP031
+                (" ", False),
+                ("...", True),
+                ("]", False),
+            )
         elif action.nargs == argparse.REMAINDER:
             # '...'
             yield "...", True
         elif action.nargs == argparse.PARSER:
             # '%s ...' % get_metavar(1)
-            yield "%s" % get_metavar(1), True  # noqa: UP031
-            yield " ...", True
+            yield from (
+                ("%s" % get_metavar(1), True),  # noqa: UP031
+                (" ", False),
+                ("...", True),
+            )
         elif action.nargs == argparse.SUPPRESS:
             # ''
             yield "", False
```

</details>

I am now going to do the following:

1. Close PR #126, referencing this new solution
2. Test the current solution (there need to be some changes because of the new logic)
3. Open a new PR from the new solution branch, which fixes both #125 and this issue.
